src/GraphQL.Tests/Bugs/Issue2387.cs

```diff
@@ -0,0 +1,157 @@
+
+        [Fact]
+        public void codefirst_parsevalue()
+        {
+            AssertQuerySuccess("query ($arg: Int) { testInput(arg:$arg) }", "{\"testInput\": \"122\"}", "{\"arg\":123}".ToInputs());
+        }
+
+        [Fact]
+        public void codefirst_nonnull_output()
+        {
+            AssertQuerySuccess("{ testNonNullOutput }", "{\"testNonNullOutput\": 124}");
+        }
+
+        [Fact]
+        public void codefirst_nonnull_parseliteral()
+        {
+            AssertQuerySuccess("{ testNonNullInput(arg:123) }", "{\"testNonNullInput\": \"122\"}");
+        }
+
+        [Fact]
+        public void codefirst_nonnull_parsevalue()
+        {
+            AssertQuerySuccess("query ($arg: Int!) { testNonNullInput(arg:$arg) }", "{\"testNonNullInput\": \"122\"}", "{\"arg\":123}".ToInputs());
+        }
+
+        [Fact]
+        public async Task schemafirst_output()
+        {
+            var schema = BuildSchemaFirst();
+            var json = await schema.ExecuteAsync(_ =>
+            {
+                _.Query = "{ testOutput }";
+                _.Root = new SchemaFirstRoot();
+            });
+            json.ShouldBeCrossPlatJson("{\"data\":{\"testOutput\": 124}}");
+        }
+
+        [Fact]
+        public async Task schemafirst_parseliteral()
+        {
+            var schema = BuildSchemaFirst();
+            var json = await schema.ExecuteAsync(_ =>
+            {
+                _.Query = "{ testInput(arg:123) }";
+                _.Root = new SchemaFirstRoot();
+            });
+            json.ShouldBeCrossPlatJson("{\"data\":{\"testInput\": \"122\"}}");
+        }
+
+        [Fact]
+        public async Task schemafirst_parsevalue()
+        {
+            var schema = BuildSchemaFirst();
+            var json = await schema.ExecuteAsync(_ =>
+            {
+                _.Query = "query ($arg: Int!) { testInput(arg:$arg) }";
+                _.Inputs = "{\"arg\":123}".ToInputs();
+            });
+            json.ShouldBeCrossPlatJson("{\"data\":{\"testInput\": \"122\"}}");
+        }
+
+        private Schema BuildSchemaFirst()
+        {
+            var typeDefs = @"
+type Query {
+  testOutput: Int!
+  testInput(arg: Int!): String!
+}";
+            var schema = GraphQL.Types.Schema.For(typeDefs, config => config.Types.Include<SchemaFirstRoot>());
+            schema.RegisterType(new MyIntGraphType());
+            return schema;
+        }
+
+        [GraphQLMetadata("Query")]
+        public class SchemaFirstRoot
+        {
+            public int TestOutput()
+            {
+                return 123;
+            }
+            
+            public string TestInput(IResolveFieldContext context)
+            {
+                return context.GetArgument<int>("arg").ToString();
+            }
+        }
+
+        public class MySchema : Schema
+        {
+            public MySchema()
+            {
+                Query = new MyQuery();
+                RegisterType(new MyIntGraphType());
```

Comment:
Change this line to `RegisterType(typeof(MyIntGraphType));` and tests will fail. See `GetSchemaTypes` . To your idea to work, you need to register specially added types/instances in the first place. And it would be nice to add a comment on this so that no one then changed the order of returned types.

Reply:
And it is also necessary to write in the documentation "in large letters" -  _to change the scalars, you need to register them first_.

Reply:
Right. We will need to be very explicit about how this is supported. Was going to update the scalar docs PR once we finish this PR.